File: CS/MultiGroupSummary/MyGridControl/MyGridViewInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Drawing;
using DevExpress.XtraGrid;
using System.Collections;
using DevExpress.XtraGrid.Columns;
using DevExpress.Data;

namespace Q354185.MyXtraGrid {
    public class MyGridViewInfo : GridViewInfo {
        public MyGridViewInfo(GridView gridView)
            : base(gridView) {
        }
        protected override int CalcGroupFooterHeight() {
            var myView = (MyGridView)View;
            if (myView.GetGroupFooterRowCount() <= 0) {
                return 0;
            }
            return base.CalcGroupFooterHeight() * myView.GetGroupFooterRowCount();
        }
        protected override int GetMaxColumnFooterCount() {
            return ((MyGridView)View).GetGroupFooterRowCount();
        }
        protected override void CalcRowCellsFooterInfo(GridRowFooterInfo fi, GridRowInfo ri) {
            AddColumnsInfo(fi.RowHandle);
            base.CalcRowCellsFooterInfo(fi, ri);
            ChangeDisplayText(fi);
            RemoveColumnsInfo();
        }
        private void AddColumnsInfo(int rowHandle) {
            rowHandle = rowHandle >= 0 ? -1 : rowHandle;
            var groupSummary = View.GetGroupSummaryValues(rowHandle);
            var myGroupSummaryList = new List<DictionaryEntry>();
            foreach (DictionaryEntry item in groupSummary) {
                myGroupSummaryList.Add(item);
            }
            var query = from item in myGroupSummaryList
                        let summary = (GridGroupSummaryItem)item.Key
                        group item by summary.ShowInGroupColumnFooter;
            foreach (var group in query) {
                var myGroup = group.ToList();
                for (var i = 1; i < myGroup.Count; i++) {
                    var sourceCInfo = ColumnsInfo[group.Key];
                    if (sourceCInfo == null) {
                        continue;
                    }
                    AddNewColumnsInfo(myGroup, i, sourceCInfo);
                }
            }
        }
        private void AddNewColumnsInfo(List<DictionaryEntry> group, int i, GridColumnInfoArgs initCellInfo) {
            var cellInfo = new GridColumnInfoArgs(initCellInfo.Column);
            cellInfo.Assign(initCellInfo);
            cellInfo.Tag = group[i].Key as GridGroupSummaryItem;
            cellInfo.Info.StartRow = i;
            ColumnsInfo.Add(cellInfo);
        }
        private void RemoveColumnsInfo() {
            var count  = ColumnsInfo.OfType<GridColumnInfoArgs>().Where(ci => ci.Tag == null).Count() ;
            while (ColumnsInfo.Count > count) {
                ColumnsInfo.RemoveAt(count);
            }
        }
        private void ChangeDisplayText(GridRowFooterInfo fi) {
            foreach (GridFooterCellInfoArgs fci in fi.Cells) {
                var rowSummaryItem = View.GetRowSummaryItem(fi.RowHandle, fci.Column);
                if (rowSummaryItem.Key == null) {
                    continue;
                }
                var item = fci.ColumnInfo.Tag as GridGroupSummaryItem;
                if (item == null) {
                    fci.Visible = ((GridGroupSummaryItem)rowSummaryItem.Key).SummaryType == SummaryItemType.None ? false : true;
                    continue;
                }
                fci.Visible = item.SummaryType == SummaryItemType.None ? false : true;
                ChangeFooterCellDisplayText(fi, fci, item);
            }
        }
        private void ChangeFooterCellDisplayText(GridRowFooterInfo fi, GridFooterCellInfoArgs fci, GridGroupSummaryItem item) {
            var dEntry = GetCustomSummaryItemValue(fi.RowHandle, fci.Column, item);
            fci.Visible = item.SummaryType == SummaryItemType.None ? false : true;
            fci.Value = dEntry.Value;
            fci.DisplayText = ((GridSummaryItem)dEntry.Key).GetDisplayText(dEntry.Value, false);
        }
        public virtual DictionaryEntry GetCustomSummaryItemValue(int rowHandle, GridColumn column, object summaryItem) {
            if (column == null) {
                return new DictionaryEntry();
            }
            var summary = View.GetGroupSummaryValues(rowHandle);
            if (summary == null) {
                return new DictionaryEntry();
            }
            foreach (DictionaryEntry dEntry in summary) {
                var item = dEntry.Key as GridGroupSummaryItem;
                if (item.ShowInGroupColumnFooter != column) {
                    continue;
                }
                if (item != summaryItem) {
                    continue;
                }
                return dEntry;
            }
            return new DictionaryEntry();
        }
        protected override void UpdateRowFooterInfo(GridRowInfo ri, GridRowFooterInfo fi) {
            base.UpdateRowFooterInfo(ri, fi);
            ChangeDisplayText(fi);
        }
    }
}

File: VB/MultiGroupSummary/MyGridControl/MyGridViewInfo.vb
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.XtraGrid
Imports System.Collections
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.Data

Namespace Q354185.MyXtraGrid

    Public Class MyGridViewInfo
        Inherits GridViewInfo

        Public Sub New(ByVal gridView As GridView)
            MyBase.New(gridView)
        End Sub

        Protected Overrides Function CalcGroupFooterHeight() As Integer
            Dim myView = CType(View, MyGridView)
            If myView.GetGroupFooterRowCount() <= 0 Then
                Return 0
            End If

            Return MyBase.CalcGroupFooterHeight() * myView.GetGroupFooterRowCount()
        End Function

        Protected Overrides Function GetMaxColumnFooterCount() As Integer
            Return CType(View, MyGridView).GetGroupFooterRowCount()
        End Function

        Protected Overrides Sub CalcRowCellsFooterInfo(ByVal fi As GridRowFooterInfo, ByVal ri As GridRowInfo)
            AddColumnsInfo(fi.RowHandle)
            MyBase.CalcRowCellsFooterInfo(fi, ri)
            ChangeDisplayText(fi)
            RemoveColumnsInfo()
        End Sub

        Private Sub AddColumnsInfo(ByVal rowHandle As Integer)
            rowHandle = If(rowHandle >= 0, -1, rowHandle)
            Dim groupSummary = View.GetGroupSummaryValues(rowHandle)
            Dim myGroupSummaryList = New List(Of DictionaryEntry)()
            For Each item As DictionaryEntry In groupSummary
                myGroupSummaryList.Add(item)
            Next

             ''' Cannot convert LocalDeclarationStatementSyntax, System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.VisualBasic.Syntax.EmptyStatementSyntax' to type 'Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax'.
'''    at ICSharpCode.CodeConverter.VB.CommonConversions.RemodelVariableDeclaration(VariableDeclarationSyntax declaration)
'''    at ICSharpCode.CodeConverter.VB.MethodBodyExecutableStatementVisitor.VisitLocalDeclarationStatement(LocalDeclarationStatementSyntax node)
'''    at Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor`1.Visit(SyntaxNode node)
'''    at ICSharpCode.CodeConverter.VB.CommentConvertingMethodBodyVisitor.DefaultVisit(SyntaxNode node)
''' 
''' Input:
'''             var query = from item in myGroupSummaryList
'''                         let summary = (DevExpress.XtraGrid.GridGroupSummaryItem)item.Key
'''                         group item by summary.ShowInGroupColumnFooter;
''' 
'''  For Each group In query
                Dim myGroup = group.ToList()
                For i = 1 To myGroup.Count - 1
                    Dim sourceCInfo = ColumnsInfo(group.Key)
                    If sourceCInfo Is Nothing Then
                        Continue For
                    End If

                    AddNewColumnsInfo(myGroup, i, sourceCInfo)
                Next
            Next
        End Sub

        Private Sub AddNewColumnsInfo(ByVal group As List(Of DictionaryEntry), ByVal i As Integer, ByVal initCellInfo As GridColumnInfoArgs)
            Dim cellInfo = New GridColumnInfoArgs(initCellInfo.Column)
            cellInfo.Assign(initCellInfo)
            cellInfo.Tag = TryCast(group(i).Key, GridGroupSummaryItem)
            cellInfo.Info.StartRow = i
            ColumnsInfo.Add(cellInfo)
        End Sub

        Private Sub RemoveColumnsInfo()
            Dim count = ColumnsInfo.OfType(Of GridColumnInfoArgs)().Where(Function(ci) ci.Tag Is Nothing).Count()
            While ColumnsInfo.Count > count
                ColumnsInfo.RemoveAt(count)
            End While
        End Sub

        Private Sub ChangeDisplayText(ByVal fi As GridRowFooterInfo)
            For Each fci As GridFooterCellInfoArgs In fi.Cells
                Dim rowSummaryItem = View.GetRowSummaryItem(fi.RowHandle, fci.Column)
                If rowSummaryItem.Key Is Nothing Then
                    Continue For
                End If

                Dim item = TryCast(fci.ColumnInfo.Tag, GridGroupSummaryItem)
                If item Is Nothing Then
                    fci.Visible = If(CType(rowSummaryItem.Key, GridGroupSummaryItem).SummaryType = SummaryItemType.None, False, True)
                    Continue For
                End If

                fci.Visible = If(item.SummaryType = SummaryItemType.None, False, True)
                ChangeFooterCellDisplayText(fi, fci, item)
            Next
        End Sub

        Private Sub ChangeFooterCellDisplayText(ByVal fi As GridRowFooterInfo, ByVal fci As GridFooterCellInfoArgs, ByVal item As GridGroupSummaryItem)
            Dim dEntry = GetCustomSummaryItemValue(fi.RowHandle, fci.Column, item)
            fci.Visible = If(item.SummaryType = SummaryItemType.None, False, True)
            fci.Value = dEntry.Value
            fci.DisplayText = CType(dEntry.Key, GridSummaryItem).GetDisplayText(dEntry.Value, False)
        End Sub

        Public Overridable Function GetCustomSummaryItemValue(ByVal rowHandle As Integer, ByVal column As GridColumn, ByVal summaryItem As Object) As DictionaryEntry
            If column Is Nothing Then
                Return New DictionaryEntry()
            End If

            Dim summary = View.GetGroupSummaryValues(rowHandle)
            If summary Is Nothing Then
                Return New DictionaryEntry()
            End If

            For Each dEntry As DictionaryEntry In summary
                Dim item = TryCast(dEntry.Key, GridGroupSummaryItem)
                If item.ShowInGroupColumnFooter IsNot column Then
                    Continue For
                End If

                If item IsNot summaryItem Then
                    Continue For
                End If

                Return dEntry
            Next

            Return New DictionaryEntry()
        End Function

        Protected Overrides Sub UpdateRowFooterInfo(ByVal ri As GridRowInfo, ByVal fi As GridRowFooterInfo)
            MyBase.UpdateRowFooterInfo(ri, fi)
            ChangeDisplayText(fi)
        End Sub
    End Class
End Namespace
